source/EventsDriven/Fido.EventsDriven.Utils.pas

```diff
@@ -0,0 +1,134 @@
+(*
+ * Copyright 2022 Mirko Bianco (email: writetomirko@gmail.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *)
+
+unit Fido.EventsDriven.Utils;
+
+interface
+
+uses
+  System.SysUtils,
+  System.TypInfo,
+  System.Generics.Collections,
+  System.JSON,
+  System.Rtti,
+  System.Math,
+
+  Spring,
+  Spring.Collections,
+
+  Fido.JSON.Marshalling,
+  Fido.Exceptions,
+  Fido.JSON.Mapping;
+
+type
+  EEventsDrivenUtilities = class(EFidoException);
+
+  TMappingMethod = reference to function(const Payload: TValue; const ParamsCount: Integer; const Method: TRttiMethod): TArray<TValue>;
+
+  TEventsDrivenUtilities = record
+  private class var
+    FMappings: IDictionary<Pointer, TMappingMethod>;
+  public
+    class constructor Create;
+    class function FormatKey(const Channel: string; const EventName: string): string; static;
+    class function PayloadToMethodParams<PayloadType>(const Payload: PayloadType; const Method: TRttiMethod): TArray<TValue>; static;
+
+    class procedure RegisterPayloadTypeMapper<PayloadType>(const Method: TMappingMethod); static;
+  end;
+
+implementation
+
+{ TEventsDrivenUtilities }
+
+class constructor TEventsDrivenUtilities.Create;
+begin
+  FMappings := TCollections.CreateDictionary<Pointer, TMappingMethod>;
+
+  FMappings.Items[TypeInfo(string)] := function(const Payload: TValue; const ParamsCount: Integer; const Method: TRttiMethod): TArray<TValue>
+    var
+      JSONValue: Shared<TJSONValue>;
+      Index: Integer;
+    begin
+      SetLength(Result, ParamsCount);
+
+      JSONValue := TJSONValue.ParseJSONValue(Payload.AsType<string>);
```

Comment:
```suggestion
      JSONValue := TJSONObject.ParseJSONValue(Payload.AsType<string>);
```

Reply:
Historically I always used TJSONValue... also on Internet it is called using TJSONValue.
And it makes sense if you are looking at it from the semantic point of view.
You are parsing a value that could be an object, an array or whatever.
Do you agree?

Reply:
It does, but on my machine (I know, I know) it keeps screaming during compiling that it doesn't work -.-
And (I must be blind) I cannot see ParseJSONValue in the System.JSON on the TJSONValue object